src/main/java/org/dataloader/DataLoaderHelper.java

```diff
@@ -546,4 +617,197 @@ private CompletableFuture<List<V>> setToValueCache(List<V> assembledValues, List
+
+            K key = keys.get(idx);
+            Object callContext = callContexts.get(idx);
+            CompletableFuture<V> future = queuedFutures.get(idx);
+            if (value instanceof Try) {
+                // we allow the batch loader to return a Try so we can better represent a computation
+                // that might have worked or not.
+                Try<V> tryValue = (Try<V>) value;
+                if (tryValue.isSuccess()) {
+                    future.complete(tryValue.get());
+                } else {
+                    stats.incrementLoadErrorCount(new IncrementLoadErrorCountStatisticsContext<>(key, callContext));
+                    future.completeExceptionally(tryValue.getThrowable());
+                    clearCacheKeys.add(keys.get(idx));
+                }
+            } else {
+                future.complete(value);
+            }
+
+            completedValues.add(value);
+            idx++;
+        }
+
+        @Override
+        public void onComplete() {
+            assertState(!onErrorCalled, () -> "onError has already been called; onComplete may not be invoked.");
+            onCompleteCalled = true;
+
+            assertResultSize(keys, completedValues);
+
+            possiblyClearCacheEntriesOnExceptions(clearCacheKeys);
+            valuesFuture.complete(completedValues);
+        }
+
+        @Override
+        public void onError(Throwable ex) {
+            assertState(!onCompleteCalled, () -> "onComplete has already been called; onError may not be invoked.");
+            onErrorCalled = true;
+
+            stats.incrementBatchLoadExceptionCount(new IncrementBatchLoadExceptionCountStatisticsContext<>(keys, callContexts));
+            if (ex instanceof CompletionException) {
+                ex = ex.getCause();
+            }
+            // Set the remaining keys to the exception.
+            for (int i = idx; i < queuedFutures.size(); i++) {
+                K key = keys.get(i);
+                CompletableFuture<V> future = queuedFutures.get(i);
+                future.completeExceptionally(ex);
+                // clear any cached view of this key because they all failed
+                dataLoader.clear(key);
+            }
+        }
+    }
+
+    private class DataLoaderMapEntrySubscriber implements Subscriber<Map.Entry<K, V>> {
+        private final CompletableFuture<List<V>> valuesFuture;
+        private final List<K> keys;
+        private final List<Object> callContexts;
+        private final List<CompletableFuture<V>> queuedFutures;
+        private final Map<K, Object> callContextByKey;
+        private final Map<K, CompletableFuture<V>> queuedFutureByKey;
+
+        private final List<K> clearCacheKeys = new ArrayList<>();
+        private final Map<K, V> completedValuesByKey = new HashMap<>();
+        private boolean onErrorCalled = false;
+        private boolean onCompleteCalled = false;
```

Comment:
```
        private final CompletableFuture<List<V>> valuesFuture;
        private final List<K> keys;
        private final List<Object> callContexts;
        private final List<CompletableFuture<V>> queuedFutures;

        private final List<K> clearCacheKeys = new ArrayList<>();
        private final List<V> completedValues = new ArrayList<>();
        private int idx = 0;

        private boolean onErrorCalled = false;
        private boolean onCompleteCalled = false;
```

90% of the variables required at the same - can we not derive from a common class and hence get some code sharing?